rotkehlchen/chain/ethereum/adex/adex.py

```diff
@@ -0,0 +1,352 @@
+    from rotkehlchen.chain.ethereum.manager import EthereumManager
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+class Adex(EthereumModule):
+    """AdEx integration module
+
+    AdEx subgraph:
+    https://github.com/samparsky/adex_subgraph
+    """
+    def __init__(
+            self,
+            ethereum_manager: 'EthereumManager',
+            database: 'DBHandler',
+            premium: Optional[Premium],
+            msg_aggregator: MessagesAggregator,
+    ) -> None:
+        self.ethereum = ethereum_manager
+        self.database = database
+        self.premium = premium
+        self.msg_aggregator = msg_aggregator
+        self.trades_lock = Semaphore()
+        try:
+            self.graph: Optional[Graph] = Graph(
+                'https://api.thegraph.com/subgraphs/name/adexnetwork/adex-protocol',
+            )
+        except RemoteError as e:
+            self.graph = None
+            self.msg_aggregator.add_error(
+                f'Could not initialize the AdEx subgraph due to {str(e)}. '
+                f'All AdEx balances and historical queries are not functioning until this is fixed. '  # noqa: E501
+                f'Probably will get fixed with time. If not report it to Rotkis support channel ',
+            )
+
+    @staticmethod
+    def _calculate_adex_balances(
+            bonds: List[Bond],
+            unbonds: List[Unbond],
+            unbond_requests: List[UnbondRequest],
+            adx_usd_price: Price,
+    ) -> Dict[ChecksumAddress, List[PoolBalance]]:
+        """Given a list of bonds, unbonds and unbond requests return the
+        balance in the protocol (staked amount) per address as <LiquidityPool>).
+
+        Given an address, its staked amount is computed by deducting the unbonds
+        and the unbond requests from its bonds.
+        """
+        address_bonds = defaultdict(list)
+        address_unbonds_set = {(unbond.address, unbond.bond_id) for unbond in unbonds}
+        address_unbond_requests_set = {
+            (unbond_request.address, unbond_request.bond_id) for unbond_request in unbond_requests
+        }
+        # Get bonds that have not been either unbonded or requested to be unbond
+        for bond in bonds:
+            if (
+                (bond.address, bond.bond_id) in
+                address_unbonds_set.union(address_unbond_requests_set)
+            ):
+                continue
+            address_bonds[bond.address].append(bond)
+
+        # Get per address staked balances in pools
+        adex_balances: DefaultDict[ChecksumAddress, List[PoolBalance]] = defaultdict(list)
+        for address, bonds in address_bonds.items():
+            pool_ids = {bond.pool_id for bond in bonds}
+            for pool_id in pool_ids:
+                amount = FVal(sum(bond.amount for bond in bonds if bond.pool_id == pool_id))
+                pool_balance = PoolBalance(
+                    pool_id=pool_id,
+                    pool_name=POOL_ID_POOL_NAME[pool_id],
```

Comment:
Probably I should improve how I get the pool name (e.g. using `get()`) and add some logs for updating the map if new pools are opened?

I need to check if LoyalityPool is a proper contract, I'm not sure. Via the subgraph I've only found Tom related events.

Reply:
It probably is. But to be sure just ask in their discord or ask Ivo.

And good idea for the logs, even though logs are only checked when a user has a problem and sends them to us.